seed/seedwriter/writer.go

```diff
@@ -0,0 +1,498 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package seedwrite implements writing image seeds.
+package seedwriter
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+// Options holds the options for a Writer.
+type Options struct {
+	RootDir string
+	SeedDir string
+
+	DefaultChannel string
+
+	// Architecture to use if none is specified by the model,
+	// useful only for classic mode. If set must match the model otherwise.
+	Architecture string
+}
+
+// OptionSnap represents options about a snap.
```

Comment:
(bikesheed) Maybe "SnapOptions" or "OptionsForSnap"? The "OptionSnap" makes it sound (to me) like this is some sort of option(al?) snap.

Reply:
I don't like any of those, this can also be extra snaps.  To me option snap and optional snaps are different things,  maybe @chipaca or @deville have suggestions.

This is a struct for snaps mentioned in options, they may be also in the model or not (extra snaps). The doc is unclear on this to be fair and I can fix that.

Also even if we decide on something else I will rename it down the line, I have lot of code piled on this atm and this is fairly central/much used.

Reply:
Not sure I can add much value here as you all write and name structs all day long (eg. is it important that the naming of a struct could conflict with the name of a snap - otherwise *OptionSnap* or *OptionsSnap* is clear to me?), but this will likely come up again so it would be good to have a pattern that everyone is happy with - it should affect *SeedSnap* too, for instance.

What about adding *Ref*, `OptionSnapRef`?

Reply:
I think OptionsSnap might be a little bit clearer overall. We do use plural modifier nouns sometimes.  SeedSnap is not affected,  there are really many options but only one seed at this level here.

Reply:
Sorry, I misunderstood what this option means. Its "Snaps-that-come-from-an-external-source" like "ubuntu-image --snaps" or the new uc20 options file. So in that light my comment does not make much sense anymore and I will mark this as resolved.